### ISSUE TITLE
[How to] Detect reference for field in methods

### ISSUE THREAD

**Issue by omuomugin** [TNG/ArchUnit#857]
I've trying to write tests like below

```
ArchRuleDefinition.fields()
  .that().inSomeCondition()
  .should()
  .notBeAccessedFromMethods().inSomeCondition()
```

or 

```
ArchRuleDefinition.methods()
  .that().inSomeCondition()
  .should()
  .notBeAccessFields().inSomeCondition()
```

Is there any kind of support for detecting dependency of fields in methods.

**Comment by codecholeric** [comment]
I think this is unfortunately still missing from the fluent API, even though it would be a nice addition! For the moment you will have to resort to add this yourself, e.g. like

```
@ArchTest
public static final ArchRule example =
  fields().that().inSomeCondition()
    .should(notBeAccessedByMethodsThat(inSomeConditionPredicate));

private static ArchCondition<JavaField> notBeAccessedByMethodsThat(DescribedPredicate<JavaMethod> predicate) {
  return new ArchCondition<JavaField>("not be accessed by methods that " + predicate.getDescription()) {
  @Override
  public void check(JavaField field, ConditionEvents events) {
    field.getAccessesToSelf().stream()
      .filter(access ->
        access.getOrigin() instanceof JavaMethod
          && predicate.test((JavaMethod) access.getOrigin())
      )
      .forEach(violation ->
        events.add(SimpleConditionEvent.violated(field, violation.getDescription()))
      );
  }
  };
}
```

I'll keep the issue open, maybe somebody wants to contribute this to the fluent API (or I'll do it once I get around to it, but no promises there about the timeline, since there are a lot of other issues I have to tackle first :grimacing:)

**Comment by omuomugin** [comment]
Thanks for your help.
I succeeded to handle it just like the example you gave me.

One little point to point out was `predicate.test` was actually `predicate.apply` (just in case for other people not to be confused)

However adding these kinds of support to fluent API will help myself so I can give it a try.

I'm thinking of API like below.

```
fields()
.that()
.inSomeCondition()
.should()
.notBeAccessedByMethods()
.that()
.inSomeCondition()
```

Was there any alternative that you were thinking of?

**Comment by codecholeric** [comment]
Oh sorry, I was playing around on the branch https://github.com/TNG/ArchUnit/pull/833 :grimacing: Once that is merged (for ArchUnit 1.0.0rc1 :wink:) the method will be `predicate.test(..)`...
Anyway, glad you could solve it :smiley:

**Comment by leonardhusmann** [comment]
Hi, am I correct in assuming that this issue is still available? If so, I would like to work on it :)
